stix2/sources/__init__.py

```diff
@@ -557,150 +687,198 @@ def add_filter(self, filters):
     def deduplicate(self, stix_obj_list):
-        """deduplicate a list of STIX objects into a unique set
+        """deduplicate a list fo STIX objects to a unique set
 
-        reduces a set of STIX objects to unique set by looking
+        Reduces a set of STIX objects to unique set by looking
         at 'id' and 'modified' fields - as a unique object version
         is determined by the combination of those fields
 
         Args:
             stix_obj_list (list): list of STIX objects (dicts)
 
         Returns:
-            (list): a unique set of the passed STIX object list
-
-
+                (list): unique set of the passed list of STIX objects
         """
+
         unique = []
-        have = False
+        dont_have = False
         for i in stix_obj_list:
+            dont_have = False
             for j in unique:
-                if i['id'] == j['id'] and i['modified'] == j['modified']:
-                    have = True
+                for field in STIX_VERSION_FIELDS:
+                    if not i[field] == j[field]:
+                        dont_have = True
                     break
-            if not have:
+            if dont_have:
                 unique.append(i)
-            have = False
         return unique
+
+
+class STIXCommonPropertyFilters():
+    """
+    """
+    @classmethod
+    def _all(cls, filter_, stix_obj_field):
+        """all filter operations (for filters whose value type can be applied to any operation type)"""
+        if filter_["op"] == '=':
+            return stix_obj_field == filter_["value"]
+        elif filter_["op"] == "!=":
+            return stix_obj_field != filter_["value"]
+        elif filter_["op"] == "in":
+            return stix_obj_field in filter_["value"]
+        elif filter_["op"] == ">":
+            return stix_obj_field > filter_["value"]
+        elif filter_["op"] == "<":
+            return stix_obj_field < filter_["value"]
+        elif filter_["op"] == ">=":
+            return stix_obj_field >= filter_["value"]
+        elif filter_["op"] == "<=":
+            return stix_obj_field <= filter_["value"]
+        else:
+            return -1
+
+    @classmethod
+    def _id(cls, filter_, stix_obj_id):
+        """base filter types"""
+        if filter_["op"] == "=":
+            return stix_obj_id == filter_["value"]
+        elif filter_["op"] == "!=":
+            return stix_obj_id != filter_["value"]
+        else:
+            return -1
+
+    @classmethod
+    def _boolean(filter_, stix_obj_field):
+        if filter_["op"] == "=":
+            return stix_obj_field == filter_["value"]
+        elif filter_["op"] == "!=":
+            return stix_obj_field != filter_["value"]
+        else:
+            return -1
+
+    @classmethod
+    def _string(cls, filter_, stix_obj_field):
+        return cls._all(filter_, stix_obj_field)
+
+    @classmethod
+    def _timestamp(cls, filter_, stix_obj_timestamp):
+        return cls._all(filter_, stix_obj_timestamp)
+
```

Comment:
I was also wondering about this. It feels tedious to me to require a separate function for each property. Could we have a function for each data type and then a mapping of property names to datatype functions? Then adding a new property would only require adding an entry to the mapping rather than defining a whole new function. Or is there a better way to handle it?